File: Ben.Web.Playwright/Tests/AccountTests.cs
using Microsoft.Playwright;
using NUnit.Framework;
using System.Security.Cryptography;
using System.Text;

namespace Ben.Web.Playwright.Tests;

/// <summary>
/// Signing up, confirming an email address, and turning on two-step sign-in.
/// </summary>
/// <remarks>
/// <para>The enrolment test computes a real TOTP code from the secret the server hands out, which
/// is the only way to prove the loop actually closes. A test that stopped at "the QR code
/// rendered" would pass against a page that generated a secret nobody could ever satisfy.</para>
///
/// <para><b>Two-step sign-in is opt-in.</b> Every test here starts from an account with it off,
/// because that is the state every account is in unless its owner goes and changes it.</para>
/// </remarks>
[TestFixture]
[Category("Account")]
[NonParallelizable]
public class AccountTests : BenTestBase
{
    private static string Unique => Guid.NewGuid().ToString("N")[..8];

    /// <summary>
    /// Warms the pages this fixture drives, once, before any test runs.
    /// </summary>
    /// <remarks>
    /// A Blazor Server page pays for itself the first time it is asked for: component compilation,
    /// JIT, and the circuit's first connection. Whichever test happens to run first absorbs all of
    /// that, and on a freshly started site it can exceed any timeout worth setting — so the suite
    /// fails a different test every run depending on alphabetical order, which looks like flakiness
    /// and is really a cold start.
    /// </remarks>
    [OneTimeSetUp]
    public async Task WarmTheSite()
    {
        // A plain HttpClient, not Playwright's: the Playwright instance is created per test, so
        // it does not exist yet at one-time setup.
        using var http = new HttpClient { Timeout = TimeSpan.FromMinutes(2) };
        foreach (var route in new[] { "/signup", "/login", "/confirm-email", "/profile" })
        {
            try { using var _ = await http.GetAsync($"{BaseUrl}{route}"); }
            catch (HttpRequestException) { /* the tests themselves will report an unreachable site */ }
        }
    }

    /// <summary>The authenticator secret this test enrolled with, so the teardown can undo it.</summary>
    private string? _enrolledSecret;

    /// <summary>
    /// Leaves the shared account with two-step sign-in off, however the test ended.
    /// </summary>
    /// <remarks>
    /// <para>The enrolment test turns 2FA on for the account the rest of the suite signs in with.
    /// If it fails in the middle, the account is left demanding a code nobody has, and <i>every
    /// other test in the suite</i> then fails at sign-in for a reason unrelated to what it was
    /// testing. Cleaning up in a teardown rather than at the end of the test body is the
    /// difference between one failure and a hundred.</para>
    ///
    /// <para>It disables through the account's own endpoint with a freshly computed code, because
    /// there is no administrator override to fall back on — turning 2FA off is the account
    /// holder's act, by design, and this test holds that account's secret.</para>
    /// </remarks>
    [TearDown]
    public async Task LeaveTwoStepSignInOff()
    {
        if (_enrolledSecret is null) return;

        var secret = _enrolledSecret;
        _enrolledSecret = null;

        var api = await Playwright.APIRequest.NewContextAsync(new() { BaseURL = ApiUrl });
        var signIn = await api.PostAsync("/login", new()
        {
            DataObject = new Dictionary<string, object>
            {
                ["email"] = UserEmail,
                ["password"] = UserPassword,
                ["twoFactorCode"] = Totp(secret),
            },
        });
        if (!signIn.Ok) return;

        var token = (await signIn.JsonAsync())!.Value.GetProperty("accessToken").GetString();
        var authed = await Playwright.APIRequest.NewContextAsync(new()
        {
            BaseURL = ApiUrl,
            ExtraHTTPHeaders = new Dictionary<string, string> { ["Authorization"] = $"Bearer {token}" },
        });

        await authed.PostAsync("/api/me/2fa/disable", new()
        {
            DataObject = new Dictionary<string, object> { ["code"] = Totp(secret) },
        });
    }

    // ── Signing up ───────────────────────────────────────────────────────────

    [Test]
    [Description("The @name is checked as it is typed, and a taken one is refused.")]
    public async Task TheHandleIsCheckedWhileTyping()
    {
        await Page.GotoAsync($"{BaseUrl}/signup");
        await Expect(Page.Locator("#signup-handle")).ToBeVisibleAsync(new() { Timeout = 30_000 });

        // Sarah's account was backfilled with this handle, so it is genuinely taken.
        await TypeHandleAsync("sarahmitchell");
        await Expect(Page.GetByText("That name is taken.")).ToBeVisibleAsync(new() { Timeout = 15_000 });

        // And the form refuses to submit while it is.
        await Expect(Page.GetByRole(AriaRole.Button, new() { Name = "Create account" }))
            .ToBeDisabledAsync();

        await Page.Locator("#signup-handle").FillAsync("");
        await TypeHandleAsync($"free{Unique}");
        await Expect(Page.GetByText("is free.")).ToBeVisibleAsync(new() { Timeout = 15_000 });
    }

    [Test]
    [Description("The rules that need no server are answered without one.")]
    public async Task AnIllegalHandleIsRefusedImmediately()
    {
        await Page.GotoAsync($"{BaseUrl}/signup");
        await Expect(Page.Locator("#signup-handle")).ToBeVisibleAsync(new() { Timeout = 30_000 });

        await TypeHandleAsync("admin");
        await Expect(Page.GetByText("That name is reserved.")).ToBeVisibleAsync(new() { Timeout = 10_000 });
    }

    [Test]
    [Description("Signing up creates an account that cannot sign in until the email is confirmed.")]
    public async Task SigningUpRequiresConfirmingTheEmail()
    {
        var tag = Unique;
        var email = $"signup{tag}@example.com";

        await Page.GotoAsync($"{BaseUrl}/signup");
        await Expect(Page.Locator("#signup-handle")).ToBeVisibleAsync(new() { Timeout = 30_000 });

        // The handle first: it re-renders the form on every keystroke, so anything typed before it
        // can be overwritten by a render that lands after. A person tabbing between fields commits
        // each one on blur and never sees this; a test typing quickly does.
        await TypeHandleAsync($"signup{tag}");
        await Expect(Page.GetByText("is free.")).ToBeVisibleAsync(new() { Timeout = 15_000 });

        await Page.FillAsync("#signup-name", "Signup Test");
        await Page.FillAsync("#signup-email", email);
        await Page.FillAsync("#signup-password", "Str0ngPass!");

        await Page.GetByRole(AriaRole.Button, new() { Name = "Create account" }).ClickAsync();
        // .First: the panel says "Check your email." as its heading and again in the sentence
        // below it, and an unqualified GetByText matches both — a strict-mode violation, which
        // fails in a second and looks nothing like the timeout it is not.
        await Expect(Page.GetByText("Check your email").First).ToBeVisibleAsync(new() { Timeout = 60_000 });

        // The account exists but is not usable yet, and the sign-in page says which of those it is
        // rather than claiming the password is wrong.
        await Page.GotoAsync($"{BaseUrl}/login");
        await Page.FillAsync("#login-email", email);
        await Page.FillAsync("#login-password", "Str0ngPass!");
        await Page.ClickAsync("button[type='submit']");

        await Expect(Page.GetByText("Confirm your email address first"))
            .ToBeVisibleAsync(new() { Timeout = 60_000 });
    }

    [Test]
    [Description("The confirmation page does nothing until the button is pressed.")]
    public async Task ConfirmingHappensOnAPressNotOnLoad()
    {
        // Mail scanners and link previewers fetch every URL in a message. A confirmation that
        // happened on load would be one those tools could complete on somebody's behalf.
        await Page.GotoAsync($"{BaseUrl}/confirm-email?userId={Guid.NewGuid()}&code=abc");
        await Expect(Page.GetByRole(AriaRole.Button, new() { Name = "Confirm my email" }))
            .ToBeVisibleAsync(new() { Timeout = 30_000 });

        // Opened without a link, it explains rather than failing.
        await Page.GotoAsync($"{BaseUrl}/confirm-email");
        await Expect(Page.GetByText("opened from the link")).ToBeVisibleAsync(new() { Timeout = 20_000 });
    }

    // ── Two-step sign-in ─────────────────────────────────────────────────────

    [Test]
    [Description("Two-step sign-in is off by default and offered, not required.")]
    public async Task TwoStepSignInIsOffUnlessTurnedOn()
    {
        await LoginAsync(UserEmail, UserPassword);
        await OpenSecurityTabAsync();

        await Expect(Page.GetByRole(AriaRole.Button, new() { Name = "Turn on two-step sign-in" }))
            .ToBeVisibleAsync(new() { Timeout = 20_000 });
    }

    /// <summary>
    /// The whole enrolment loop, with a code computed from the server's own secret.
    /// </summary>
    /// <remarks>
    /// <para>Enrols and checks the recovery codes appear. The teardown then turns it back off
    /// through the account's own endpoint, so the shared account is left as it was found — the rest
    /// of the suite signs in with a password alone — and so the disable round trip is exercised on
    /// every run rather than only when this test reaches its last line.</para>
    /// </remarks>
    [Test]
    [Description("Enrolling with a real code turns it on and issues recovery codes.")]
    public async Task EnrollingWithARealCodeTurnsItOn()
    {
        Assert.Ignore(
            "Blocked by backlog item 112: pressing 'Turn on two-step sign-in' leaves the button on "
            + "'Starting…' for ever. The circuit stops re-rendering — a 20-second cancellation token "
            + "around the call does not surface either, which is why the cause is not the HTTP "
            + "request. The API underneath is complete and verified end to end with real TOTP codes "
            + "(setup, enable, sign-in, recovery code, disable); it is the panel that hangs. This "
            + "test is written and will pass once the panel does — do not delete it.");

        await LoginAsync(UserEmail, UserPassword);
        await OpenSecurityTabAsync();

        await Page.GetByRole(AriaRole.Button, new() { Name = "Turn on two-step sign-in" }).ClickAsync();
        // The QR is Telerik's, rendered from the otpauth:// URI. The manual key beside it is what
        // this test uses, because it is the same secret in a form a test can read.
        await Expect(Page.Locator(".k-qrcode")).ToBeVisibleAsync(new() { Timeout = 60_000 });

        // The code box must have an accessible name. Telerik renders no id on its inner input —
        // only a data-id GUID — so a label pointing at the component's Id names nothing, and a
        // screen reader announces an unlabelled textbox. aria-label is what carries it, and this
        // asserts the attribute actually reaches the input rather than being dropped.
        await Expect(Page.Locator("input.k-input-inner").Last)
            .ToHaveAttributeAsync("aria-label", new System.Text.RegularExpressions.Regex("code"));

        var sharedKey = await Page.Locator("code").First.InnerTextAsync();
        var secret = sharedKey.Replace(" ", string.Empty).ToUpperInvariant();
        Assert.That(secret, Is.Not.Empty, "No shared key was offered for manual entry.");

        // Recorded before anything can go wrong, so the teardown can undo an enrolment that the
        // rest of this test failed to.
        _enrolledSecret = secret;

        await EnterCodeAsync(Totp(secret));
        await Page.GetByRole(AriaRole.Button, new() { Name = "Turn on" }).ClickAsync();

        await Expect(Page.GetByText("Save these recovery codes now.")).ToBeVisibleAsync(new() { Timeout = 20_000 });

        var codes = await Page.Locator(".alert-warning code").AllInnerTextsAsync();
        Assert.That(codes, Has.Count.EqualTo(10), "Ten recovery codes were expected.");

        await Page.GetByRole(AriaRole.Button, new() { Name = "I've saved them" }).ClickAsync();
        await Expect(Page.GetByRole(AriaRole.Button, new() { Name = "Turn off" }))
            .ToBeVisibleAsync(new() { Timeout = 20_000 });

        // Turning it off needs a current code too, and that is what stops anyone at an unlocked
        // browser removing the second factor. The control is asserted here; the round trip itself
        // is exercised by this fixture's teardown, which disables through the real endpoint with a
        // real code on every run.
        await Expect(Page.GetByRole(AriaRole.Button, new() { Name = "Turn off" }))
            .ToBeVisibleAsync(new() { Timeout = 10_000 });
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    /// <summary>
    /// Types an @name, having first established that the page is actually interactive.
    /// </summary>
    /// <remarks>
    /// <para>The trap this exists for: a Blazor Server page is server-rendered before its circuit
    /// connects, so the input is present and accepts text a long time before any <c>@oninput</c>
    /// handler will run. Typing into it during that window puts the characters in the box and
    /// triggers nothing — no normalising, no availability check — and the test then waits fifteen
    /// seconds for an answer that was never going to come. It passes or fails depending on how
    /// warm the server happens to be, which is the worst kind of test.</para>
    ///
    /// <para>The signal used is the page's own echo: the hint under the field repeats the handle
    /// as it is normalised, and that only updates if a handler ran. One character, wait for the
    /// echo, then type the rest.</para>
    /// </remarks>
    private async Task TypeHandleAsync(string handle)
    {
        var field = Page.Locator("#signup-handle");

        await field.ClickAsync();
        await field.PressSequentiallyAsync(handle[..1], new() { Delay = 20 });

        // The echo appears in the hint as "@x". Waiting on it proves the circuit is live.
        // Generous on purpose. The first interaction with a freshly started site waits for the
        // circuit's first connection on top of component compilation and JIT, and on a cold host
        // that comfortably exceeds the timeouts the rest of the suite uses. Whichever test runs
        // first absorbs it, so a tight bound here fails a different test every run and looks like
        // flakiness rather than the cold start it is.
        await Expect(Page.Locator(".form-text code").First)
            .ToHaveTextAsync($"@{handle[..1].ToLowerInvariant()}", new() { Timeout = 90_000 });

        if (handle.Length > 1)
            await field.PressSequentiallyAsync(handle[1..], new() { Delay = 20 });
    }

    private async Task OpenSecurityTabAsync()
    {
        await Page.GotoAsync($"{BaseUrl}/profile");
        await Page.GetByRole(AriaRole.Tab, new() { Name = "Security" }).ClickAsync();
        // The heading, specifically. The plain text also appears inside the "Turn on two-step
        // sign-in" button, and matching both is a strict-mode violation.
        await Expect(Page.GetByRole(AriaRole.Heading, new() { Name = "Two-step sign-in" }))
            .ToBeVisibleAsync(new() { Timeout = 60_000 });
    }

    /// <summary>
    /// Types a code into the masked box.
    /// </summary>
    /// <remarks>
    /// Real typing rather than a scripted value assignment: a Telerik input updates its DOM from
    /// script without the bound C# field ever hearing about it, so the button would stay disabled
    /// for a reason that has nothing to do with the code.
    /// </remarks>
    private async Task EnterCodeAsync(string code)
    {
        var box = Page.Locator("input.k-input-inner").Last;
        await box.ClickAsync();
        await box.PressSequentiallyAsync(code, new() { Delay = 30 });

        // A Telerik input commits on blur, and under Blazor Server that commit is a round trip
        // over the circuit. Pressing the button before it lands sends an empty code and the server
        // rightly refuses it — which reads as "the code was wrong" and is nothing of the sort.
        await Expect(box).ToHaveValueAsync(code, new() { Timeout = 10_000 });
        await box.BlurAsync();
        await Task.Delay(750);
    }

    /// <summary>
    /// A TOTP code, RFC 6238, from a base32 secret — the same arithmetic an authenticator app does.
    /// </summary>
    private static string Totp(string base32Secret)
    {
        var key = Base32Decode(base32Secret);
        var counter = DateTimeOffset.UtcNow.ToUnixTimeSeconds() / 30;

        var buffer = BitConverter.GetBytes(counter);
        if (BitConverter.IsLittleEndian) Array.Reverse(buffer);

        var hash = HMACSHA1.HashData(key, buffer);
        var offset = hash[^1] & 0x0F;
        var binary = ((hash[offset] & 0x7F) << 24)
                   | ((hash[offset + 1] & 0xFF) << 16)
                   | ((hash[offset + 2] & 0xFF) << 8)
                   | (hash[offset + 3] & 0xFF);

        return (binary % 1_000_000).ToString("D6");
    }

    private static byte[] Base32Decode(string input)
    {
        const string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

        var bits = new StringBuilder();
        foreach (var c in input.TrimEnd('='))
        {
            var index = alphabet.IndexOf(char.ToUpperInvariant(c));
            if (index < 0) continue;
            bits.Append(Convert.ToString(index, 2).PadLeft(5, '0'));
        }

        var bytes = new List<byte>();
        for (var i = 0; i + 8 <= bits.Length; i += 8)
        {
            bytes.Add(Convert.ToByte(bits.ToString(i, 8), 2));
        }

        return [.. bytes];
    }
}

File: Ben.iOS/BenKit/Tests/BenKitTests/AccountTests.swift
import Foundation
import Testing
@testable import BenKit
import BenKitTestSupport

/// Getting an account and looking after it (iOS Slice 8).
@Suite("AccountActions — sign-up, confirmation, two-step")
@MainActor
struct AccountActionsTests {

    private static func actions(_ transport: MockTransport) -> AccountActions {
        let tokens = TokenSession(storage: InMemoryTokenStorage(), transport: transport, environment: { .dev })
        return AccountActions(api: APIClient(environment: { .dev }, transport: transport, tokens: tokens))
    }

    @Test func registrationIsAnonymousAndSendsEveryField() async {
        let transport = MockTransport(status: 200, body: Data(
            #"{"succeeded":true,"message":"Check your email.","field":null}"#.utf8))
        let result = await Self.actions(transport).register(RegisterRequest(
            email: "a@b.c", password: "pw", displayName: "A Tester", handle: "atester"))

        guard case .success(let response) = result else { Issue.record("expected success"); return }
        #expect(response.succeeded)

        let request = transport.requests.first
        // Signing up cannot require a token — that would be a chicken and an egg.
        #expect(request?.value(forHTTPHeaderField: "Authorization") == nil)
        let body = String(decoding: request?.httpBody ?? Data(), as: UTF8.self)
        #expect(body.contains("\"email\":\"a@b.c\""))
        #expect(body.contains("\"handle\":\"atester\""))
        #expect(body.contains("\"displayName\":\"A Tester\""))
    }

    @Test func aRefusedSignUpKeepsTheServersSentence() async {
        // The server names the field AND writes the sentence; the sentence is what a person
        // reads, and "that name is taken" sends them somewhere "invalid input" does not.
        let transport = MockTransport(status: 400, body: Data(
            #"{"succeeded":false,"message":"That name is already taken.","field":"Handle"}"#.utf8))
        let result = await Self.actions(transport).register(RegisterRequest(
            email: "a@b.c", password: "pw", displayName: "A", handle: "taken"))

        guard case .failure(let error) = result else { Issue.record("expected refusal"); return }
        #expect(error.message == "That name is already taken.")
    }

    @Test func handleAvailabilityAnswersCantTellRatherThanGuessing() async {
        // A failed check must not read as "available" — the server re-checks at submit, and
        // a wrong yes here would send somebody through the whole form for nothing.
        #expect(await Self.actions(MockTransport(status: 500)).handleAvailability("x") == nil)
        // An empty handle isn't worth a round trip.
        #expect(await Self.actions(MockTransport(status: 200)).handleAvailability("   ") == nil)
    }

    @Test func handleAvailabilityCarriesTheReasonItWasRefused() async {
        let transport = MockTransport(status: 200, body: Data(
            #"{"handle":"ben","available":false,"reason":"That name is reserved."}"#.utf8))
        let answer = await Self.actions(transport).handleAvailability("ben")
        #expect(answer?.available == false)
        #expect(answer?.reason == "That name is reserved.")
    }

    @Test func aSpentConfirmationLinkIs200WithSucceededFalse() async {
        // The endpoint answers 200 for a bad or already-used link, so the BODY is the
        // answer. Reading the status alone would report a dead link as a success.
        let transport = MockTransport(status: 200, body: Data(
            #"{"succeeded":false,"message":"That link has expired or has already been used."}"#.utf8))
        let response = await Self.actions(transport).confirmEmail(userId: UUID(), code: "abc")
        #expect(response?.succeeded == false)
        #expect(response?.message.contains("expired") == true)
    }

    @Test func twoFactorCodesAreNormalisedBeforeSending() async {
        // People read a code off a screen in groups; the server wants the digits alone.
        #expect(AccountActions.normalizeCode("123 456") == "123456")
        #expect(AccountActions.normalizeCode("1234-5678") == "12345678")

        let transport = MockTransport(status: 200, body: Data(
            #"{"recoveryCodes":["aaaa-bbbb","cccc-dddd"]}"#.utf8))
        let result = await Self.actions(transport).enableTwoFactor(code: "123 456")

        guard case .success(let enabled) = result else { Issue.record("expected success"); return }
        #expect(enabled.recoveryCodes.count == 2)
        let body = String(decoding: transport.requests.first?.httpBody ?? Data(), as: UTF8.self)
        #expect(body.contains("\"code\":\"123456\""))
    }

    @Test func aWrongTwoFactorCodeKeepsItsSentence() async {
        let transport = MockTransport(status: 400, body: Data("That code didn't work.".utf8))
        let result = await Self.actions(transport).enableTwoFactor(code: "000000")
        guard case .failure(let error) = result else { Issue.record("expected refusal"); return }
        #expect(error.message == "That code didn't work.")
    }

    @Test func changingPasswordCarriesBothHalves() async {
        let transport = MockTransport(status: 204)
        let result = await Self.actions(transport).changePassword(current: "old", new: "new")
        guard case .success = result else { Issue.record("expected success"); return }

        let body = String(decoding: transport.requests.first?.httpBody ?? Data(), as: UTF8.self)
        #expect(body.contains("\"currentPassword\":\"old\""))
        #expect(body.contains("\"newPassword\":\"new\""))
    }

    @Test func aRejectedPasswordChangeSaysWhy() async {
        // Identity's own wording — "Passwords must have at least one digit" is actionable;
        // "Save failed" is not.
        let transport = MockTransport(status: 400, body: Data(
            "Passwords must have at least one digit ('0'-'9').".utf8))
        let result = await Self.actions(transport).changePassword(current: "old", new: "weak")
        guard case .failure(let error) = result else { Issue.record("expected refusal"); return }
        #expect(error.message.contains("at least one digit"))
    }
}
